### PR TITLE
Add Constrintern.interp_constr_evars_impls_gen taking pretyping flags

### PR DESCRIPTION
**Pull request by SkySkimmer**
And use it in various places.

Also add final solve_remaining_evars to various commands, fixing #11576.

### CODE REVIEW

File: vernac/record.ml [hunk 1]
```diff
@@ -180,7 +190,8 @@ let typecheck_params_and_fields def poly pl ps records =
   in
   let (sigma, data) = List.fold_left2_map fold sigma records arities in
   let sigma =
-    Pretyping.solve_remaining_evars Pretyping.all_and_fail_flags env_ar sigma in
+    Pretyping.solve_remaining_evars Pretyping.all_and_fail_flags env_ar sigma
+  in
```
Comment (herbelin) [inline vernac/record.ml:194]: :smile:

**Review by herbelin — APPROVED** [review]
Good to me.

Maybe the pure generalization to flags could be a separate commit from what add a final call to heuristics, but if complicated to do that's not so important either.

Don't know if other wants to have a look. Ping me if you need an assignee.

File: vernac/comDefinition.ml [hunk 2]
```diff
@@ -95,17 +95,26 @@ let interp_definition ~program_mode env evd impl_env bl red_option c ctypopt =
...
+        interp_constr_evars_impls_gen ~flags ~impls env_bl evd (Pretyping.OfType ty) c
+      in
       check_imps ~impsty ~impsbody;
       evd, c, imps1@impsty, Some ty
   in
+  (* solve typeclasses / heuristics *)
+  (* even with program_mode=false we don't fail here as we don't want
+     to fail if there are unused evars *)
+  let solve_flags = Pretyping.{all_no_fail_flags with program_mode} in
+  let evd = Pretyping.solve_remaining_evars solve_flags env evd in
```
Comment (SkySkimmer) [inline vernac/comDefinition.ml:113]: @herbelin I'm not sure about using interp_definition with this added solve in record (now that #13166 is merged), WDYT?
Maybe separate a interp_definition_core with about this https://github.com/coq/coq/blob/031160d00a8e2cfcd74709a487be260378bad0f2/vernac/comDefinition.ml#L87-L108 ?

Comment (herbelin) [reply vernac/comDefinition.ml:113]: > @herbelin I'm not sure about using interp_definition with this added solve in record (now that #13166 is merged), WDYT?

Indeed, the solve would come too early.

> Maybe separate a interp_definition_core with about this 

Or `interp_definition_component`? Or keep the `interp_definition` name and add a `interp_standalone_definition`/`interp_complete_definition` with the solve.

**Comment by ppedrot** [comment]
@SkySkimmer What's the status of this?

**Comment by SkySkimmer** [comment]
I lack the interest to finish this.

**Pull request closed without merging**